src/dune_engine/build.ml

```diff
@@ -353,79 +359,116 @@ let fold_labeled (type acc) t ~(init : acc) ~f =
     | Filter_existing_files p -> loop p acc
     | Memo m -> loop m.t acc
     | Catch (t, _) -> loop t acc
+    | Fiber _ -> acc
+    | Dyn_fiber b -> loop b acc
+    | Build b -> loop b acc
   in
   loop t init
 
 (* Execution *)
 
-module rec Execution : sig
-  val exec : 'a t -> 'a * Dep.Set.t
-end = struct
-  module Function = struct
-    type 'a input = 'a memo
-
-    type 'a output = 'a * Dep.Set.t
-
-    let name = "exec-memo"
-
-    let id m = m.id
+let do_not_use_stage_fiber f = Fiber f
+let do_not_use_stage_dyn_fiber f = Dyn_fiber f
+let do_not_use_stage_build f = Build f
+
+module Make_exec (Build_deps:sig val build_deps: Dep.Set.t -> unit Fiber.t end) = struct
+
+  module rec Execution : sig
+    val exec : 'a t -> ('a * Dep.Set.t) Fiber.t
+  end = struct
+    module Function = struct
+      type 'a input = 'a memo
+
+      type 'a output = 'a * Dep.Set.t
+
+      let name = "exec-memo"
+
+      let id m = m.id
+
+      let to_dyn m = Dyn.String m.name
+
+      let eval m = Execution.exec m.t
+    end
+
+    module Memo = Memo.Poly.Async (Function)
+
+    let exec t =
+      let open Fiber.O in
+      let file_exists = Fdecl.get file_exists_fdecl in
+      let eval_pred = Fdecl.get eval_pred_fdecl in
+      let rec go : type a. a t -> (a * Dep.Set.t) Fiber.t =
+        fun t ->
+          match t with
+          | Pure x -> Fiber.return (x, Dep.Set.empty)
+          | Map (f, a) ->
+            let+ a, dyn_deps_a = go a in
+            (f a, dyn_deps_a)
+          | Both (a, b) ->
+            let* a, dyn_deps_a = go a in
+            let+ b, dyn_deps_b = go b in
+            ((a, b), Dep.Set.union dyn_deps_a dyn_deps_b)
+          | Seq (a, b) ->
+            let* (), dyn_deps_a = go a in
+            let+ b, dyn_deps_b = go b in
+            (b, Dep.Set.union dyn_deps_a dyn_deps_b)
+          | Map2 (f, a, b) ->
+            let* a, dyn_deps_a = go a in
+            let+ b, dyn_deps_b = go b in
+            (f a b, Dep.Set.union dyn_deps_a dyn_deps_b)
+          | Deps _ -> Fiber.return ((), Dep.Set.empty)
+          | Paths_for_rule _ -> Fiber.return ((), Dep.Set.empty)
+          | Paths_glob g -> Fiber.return ((eval_pred g : Path.Set.t), Dep.Set.empty)
+          | Contents p -> Fiber.return (Io.read_file p, Dep.Set.empty)
```

Comment:
I wonder if it makes sense for these `Io.` calls to become asynchronous. It feels strange to do a synchronous call and then `Fiber.return` the result.

Reply:
(Or will that mess up the order in which we see errors?)

Reply:
Thanks! Let's keep these calls synchronous. It's unclear if we'll win anything by switching to asynchronous file reads.